baseplate/lib/circuit_breaker/factory.py

```diff
@@ -0,0 +1,164 @@
+            self.span.incr_tag(f"breakers.{breaker.name}.request.fail_fast")
+            raise BreakerTrippedError()
+
+        success: bool = True
+        try:
+            # yield to the application code that will use
+            # the client covered by this breaker. if this
+            # raises an exception we will catch it here.
+            yield self._client
+        except breakable_exceptions:
+            # only known exceptions in `breakable_exceptions` should trigger
+            # opening the circuit. the client call may raise exceptions that
+            # are a meaningful response, like defined thrift IDL exceptions.
+            success = False
+            raise
+        finally:
+            prev = breaker.state
+            breaker.register_attempt(success)
+            final = breaker.state
+            if prev != final:
+                self.span.incr_tag(
+                    f"breakers.{breaker.name}.state_change.{prev.value}.{final.value}"
+                )
+
+
+class CircuitBreakerBox:
+    """Container for a client's `Breaker`s.
+
+    Will lazily create `Breaker`s for each operation as needed. There
+    is no global coordination across operations--each `Breaker` is
+    isolated and does not consider the state or failure rates in other
+    `Breaker`s.
+
+    :param name: The base `Breaker` name. The full name is like "name.operation".
+    :param samples: See `Breaker`
+    :param trip_failure_ratio: See `Breaker`
+    :param trip_for: See `Breaker`
+    :param fuzz_ratio: See `Breaker`
+    """
+
+    def __init__(
+        self,
+        name: str,
+        samples: int,
+        trip_failure_ratio: float,
+        trip_for: timedelta,
+        fuzz_ratio: float,
+    ):
+        self.name = name
+        self.samples = samples
+        self.trip_failure_ratio = trip_failure_ratio
+        self.trip_for = trip_for
+        self.fuzz_ratio = fuzz_ratio
+        self.breaker_box: Dict[str, Breaker] = {}
+
+    def get(self, operation: str) -> Breaker:
+        # lazy add breaker into breaker box
+        if operation not in self.breaker_box:
+            breaker = Breaker(
+                name=f"{self.name}.{operation}",
+                samples=self.samples,
+                trip_failure_ratio=self.trip_failure_ratio,
+                trip_for=self.trip_for,
+                fuzz_ratio=self.fuzz_ratio,
+            )
+            self.breaker_box[operation] = breaker
+        return self.breaker_box[operation]
+
+
+def breaker_box_from_config(
+    app_config: config.RawConfig, name: str, prefix: str = "breaker.",
+) -> CircuitBreakerBox:
+    """Make a CircuitBreakerBox from a configuration dictionary."""
+    # TODO: fix default handling here. if these are not set
+    # they will be None and passed through to the Breaker() constructor
+    # which will override the defaults set in Breaker()
+    assert prefix.endswith(".")
+    parser = config.SpecParser(
```

Comment:
what is difference between `config.parse_config` vs `config.SpecParser`

Reply:
https://github.com/reddit/baseplate.py/blob/develop/baseplate/lib/config.py#L625-L635

`parse_config` is essentially `SpecParser` (returned by `Parser.from_spec`) and then `parser.parse("", app_config)`, so it seems it doesn't let you specify the prefix.

Reply:
but in the parser.parse{"", {prefix: }} can set the `prefix` in the dict right